log_cache_async.go

```diff
@@ -0,0 +1,405 @@
+		c.state.Lock()
+		c.deliverCompletionLocked(lwc)
+		c.state.Unlock()
+
+		close(syncReq.doneCh)
+
+		// Loop around, if more writes have happened, triggerChan will already have
+		// a syncReq buffered and we'll immediately start flushing again.
+	}
+}
+
+func (c *LogCacheAsync) deliverCompletionLocked(lwc *LogWriteCompletion) {
+	if lwc == nil {
+		return
+	}
+	if lwc.Error == nil {
+		atomic.StoreUint64(&c.persistentIndex, lwc.PersistentIndex)
+	}
+	c.state.completionCh <- *lwc
+}
+
+func (c *LogCacheAsync) doFlush(logs []*Log, start time.Time) *LogWriteCompletion {
+	if len(logs) < 1 {
+		return nil
+	}
+
+	// Write to the underlying store
+	err := c.store.StoreLogs(logs)
+
+	lwc := LogWriteCompletion{
+		PersistentIndex: logs[0].Index - 1,
+		Error:           err,
+		Duration:        time.Since(start),
+	}
+	if err == nil {
+		lwc.PersistentIndex = logs[len(logs)-1].Index
+	}
+	return &lwc
+}
+
+// DisableAsync implements AsyncLogStore
+func (c *LogCacheAsync) DisableAsync() {
+	c.state.Lock()
+	// Check we are in sync mode to enforce correct state machine usage.
+	if c.state.completionCh == nil {
+		panic("already in sync mode")
+	}
+	// unlock again since the flusher will need the lock to stop and clear state.
+	// This method can only be called from the serial leader loop so we don't have
+	// to worry about races from other callers of Append, Enable* or Disable*.
+	c.state.Unlock()
+
+	// First, wait for any pending writes to flush. In this case we do wait even
+	// if the buffer is ful since we need to see the ack.
+	doneCh := make(chan struct{})
+	c.state.triggerChan <- syncRequest{
+		startTime: time.Now(),
+		stop:      true, // Tell the flusher to exit and cleanup async state
+		doneCh:    doneCh,
+	}
+	// Wait for the sync to be done
+	<-doneCh
+}
+
+// StoreLogsAsync implements AsyncLogStore
+func (c *LogCacheAsync) StoreLogsAsync(logs []*Log) error {
+	c.state.Lock()
+	defer c.state.Unlock()
+
+	if c.state.completionCh == nil {
+		return errors.New("call to StoreLogsAsync when in sync mode")
+	}
+
+	start := time.Now()
+
+	persistedIdx := atomic.LoadUint64(&c.persistentIndex)
+	lastIdx := atomic.LoadUint64(&c.lastIndex)
+
+	// Make sure there is room in the cache for all the logs we need to write.
+	// It's very unlikely there won't be, but if we are writing really fast into a
+	// small cache and the flusher is blocked on IO for a while then we need to
+	// ensure we don't overwrite cache entries that are not persistent yet!
+	for !hasSpaceFor(len(logs), lastIdx, persistedIdx, c.size) {
+		// We need to block and wait for they sync thread to persist some more
+		// stuff! We do that by sending a sync request even though it's already busy
+		// this lets us get notified about when it's free. Note that even though we
+		// unlock and it's _possible_ for another StoreLogsAsync call to be made,
+		doneCh := make(chan struct{})
+		c.state.triggerChan <- syncRequest{startTime: start, doneCh: doneCh}
```

Comment:
This causes a deadlock right now since the flusher thread is likely flushing right now and then will need to take state lock before it can update the cache state and continue it's loop to receive this trigger.

We need to unlock the state lock before this send not just before the `doneCh` read